File: Supplementary_webpages/static/gallery_res/A Morphable Model For The Synthesis of 3D Faces Eq. 5/morphable_model_5.cpp
/*
E = 1/`σ_N`²`E_I` + ∑_j α_j²/`σ_S`_j² + ∑_j β_j²/`σ_T`_j²  + ∑_j (ρ_j-ρ̄_j)²/`σ_ρ`_j²

where

`σ_N` ∈ ℝ 
`E_I` ∈ ℝ
α_i ∈ ℝ
β_i ∈ ℝ
`σ_S`_i ∈ ℝ 
`σ_T`_i ∈ ℝ 
ρ_j ∈ ℝ 
ρ̄_j ∈ ℝ 
`σ_ρ`_j ∈ ℝ 
ā_i ∈ ℝ 
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

struct morphable_model_5ResultType {
    double E;
    morphable_model_5ResultType(const double & E)
    : E(E)
    {}
};

morphable_model_5ResultType morphable_model_5(
    const double & σ_N,
    const double & E_I,
    const std::vector<double> & α,
    const std::vector<double> & β,
    const std::vector<double> & σ_S,
    const std::vector<double> & σ_T,
    const std::vector<double> & ρ,
    const std::vector<double> & ρ̄,
    const std::vector<double> & σ_ρ,
    const std::vector<double> & ā)
{
    const long dim_0 = α.size();
    const long dim_1 = ρ.size();
    assert( β.size() == dim_0 );
    assert( σ_S.size() == dim_0 );
    assert( σ_T.size() == dim_0 );
    assert( ρ̄.size() == dim_1 );
    assert( σ_ρ.size() == dim_1 );
    assert( ā.size() == dim_0 );

    double sum_0 = 0;
    for(int j=1; j<=α.size(); j++){
        sum_0 += pow(α.at(j-1), 2) / double(pow(σ_S.at(j-1), 2));
    }
    double sum_1 = 0;
    for(int j=1; j<=β.size(); j++){
        sum_1 += pow(β.at(j-1), 2) / double(pow(σ_T.at(j-1), 2));
    }
    double sum_2 = 0;
    for(int j=1; j<=ρ.size(); j++){
        sum_2 += pow((ρ.at(j-1) - ρ̄.at(j-1)), 2) / double(pow(σ_ρ.at(j-1), 2));
    }
    double E = 1 / double(pow(σ_N, 2)) * E_I + sum_0 + sum_1 + sum_2;

    return morphable_model_5ResultType(E);
}


void generateRandomData(double & σ_N,
    double & E_I,
    std::vector<double> & α,
    std::vector<double> & β,
    std::vector<double> & σ_S,
    std::vector<double> & σ_T,
    std::vector<double> & ρ,
    std::vector<double> & ρ̄,
    std::vector<double> & σ_ρ,
    std::vector<double> & ā)
{
    σ_N = rand() % 10;
    E_I = rand() % 10;
    const int dim_0 = rand()%10;
    const int dim_1 = rand()%10;
    α.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        α[i] = rand() % 10;
    }
    β.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        β[i] = rand() % 10;
    }
    σ_S.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        σ_S[i] = rand() % 10;
    }
    σ_T.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        σ_T[i] = rand() % 10;
    }
    ρ.resize(dim_1);
    for(int i=0; i<dim_1; i++){
        ρ[i] = rand() % 10;
    }
    ρ̄.resize(dim_1);
    for(int i=0; i<dim_1; i++){
        ρ̄[i] = rand() % 10;
    }
    σ_ρ.resize(dim_1);
    for(int i=0; i<dim_1; i++){
        σ_ρ[i] = rand() % 10;
    }
    ā.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        ā[i] = rand() % 10;
    }
}


int main(int argc, char *argv[])
{
    srand((int)time(NULL));
    double σ_N;
    double E_I;
    std::vector<double> α;
    std::vector<double> β;
    std::vector<double> σ_S;
    std::vector<double> σ_T;
    std::vector<double> ρ;
    std::vector<double> ρ̄;
    std::vector<double> σ_ρ;
    std::vector<double> ā;
    generateRandomData(σ_N, E_I, α, β, σ_S, σ_T, ρ, ρ̄, σ_ρ, ā);
    morphable_model_5ResultType func_value = morphable_model_5(σ_N, E_I, α, β, σ_S, σ_T, ρ, ρ̄, σ_ρ, ā);
    std::cout<<"return value:\n"<<func_value.E<<std::endl;
    return 0;
}

File: Supplementary_webpages/static/gallery_res/A Morphable Model For The Synthesis of 3D Faces Eq. 5/morphable_model_5.py
"""
E = 1/`σ_N`²`E_I` + ∑_j α_j²/`σ_S`_j² + ∑_j β_j²/`σ_T`_j²  + ∑_j (ρ_j-ρ̄_j)²/`σ_ρ`_j²

where

`σ_N` ∈ ℝ 
`E_I` ∈ ℝ
α_i ∈ ℝ
β_i ∈ ℝ
`σ_S`_i ∈ ℝ 
`σ_T`_i ∈ ℝ 
ρ_j ∈ ℝ 
ρ̄_j ∈ ℝ 
`σ_ρ`_j ∈ ℝ 
ā_i ∈ ℝ 
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


class morphable_model_5ResultType:
    def __init__( self, E):
        self.E = E


def morphable_model_5(σ_N, E_I, α, β, σ_S, σ_T, ρ, ρ̄, σ_ρ, ā):
    α = np.asarray(α, dtype=np.float64)
    β = np.asarray(β, dtype=np.float64)
    σ_S = np.asarray(σ_S, dtype=np.float64)
    σ_T = np.asarray(σ_T, dtype=np.float64)
    ρ = np.asarray(ρ, dtype=np.float64)
    ρ̄ = np.asarray(ρ̄, dtype=np.float64)
    σ_ρ = np.asarray(σ_ρ, dtype=np.float64)
    ā = np.asarray(ā, dtype=np.float64)

    dim_0 = α.shape[0]
    dim_1 = ρ.shape[0]
    assert np.ndim(σ_N) == 0
    assert np.ndim(E_I) == 0
    assert α.shape == (dim_0,)
    assert β.shape == (dim_0,)
    assert σ_S.shape == (dim_0,)
    assert σ_T.shape == (dim_0,)
    assert ρ.shape == (dim_1,)
    assert ρ̄.shape == (dim_1,)
    assert σ_ρ.shape == (dim_1,)
    assert ā.shape == (dim_0,)

    sum_0 = 0
    for j in range(1, len(α)+1):
        sum_0 += np.power(α[j-1], 2) / np.power(σ_S[j-1], 2)
    sum_1 = 0
    for j in range(1, len(β)+1):
        sum_1 += np.power(β[j-1], 2) / np.power(σ_T[j-1], 2)
    sum_2 = 0
    for j in range(1, len(ρ)+1):
        sum_2 += np.power((ρ[j-1] - ρ̄[j-1]), 2) / np.power(σ_ρ[j-1], 2)
    E = 1 / np.power(σ_N, 2) * E_I + sum_0 + sum_1 + sum_2
    return morphable_model_5ResultType(E)


def generateRandomData():
    σ_N = np.random.randn()
    E_I = np.random.randn()
    dim_0 = np.random.randint(10)
    dim_1 = np.random.randint(10)
    α = np.random.randn(dim_0)
    β = np.random.randn(dim_0)
    σ_S = np.random.randn(dim_0)
    σ_T = np.random.randn(dim_0)
    ρ = np.random.randn(dim_1)
    ρ̄ = np.random.randn(dim_1)
    σ_ρ = np.random.randn(dim_1)
    ā = np.random.randn(dim_0)
    return σ_N, E_I, α, β, σ_S, σ_T, ρ, ρ̄, σ_ρ, ā


if __name__ == '__main__':
    σ_N, E_I, α, β, σ_S, σ_T, ρ, ρ̄, σ_ρ, ā = generateRandomData()
    print("σ_N:", σ_N)
    print("E_I:", E_I)
    print("α:", α)
    print("β:", β)
    print("σ_S:", σ_S)
    print("σ_T:", σ_T)
    print("ρ:", ρ)
    print("ρ̄:", ρ̄)
    print("σ_ρ:", σ_ρ)
    print("ā:", ā)
    func_value = morphable_model_5(σ_N, E_I, α, β, σ_S, σ_T, ρ, ρ̄, σ_ρ, ā)
    print("return value: ", func_value.E)